### ISSUE TITLE
Questions about continued pretraining DINOv2 and feature extraction in lightly_train

### ISSUE THREAD

**Issue by wudier483** [lightly-ai/lightly-train#821]
Hi, I have a few short questions:
1. Data scale for continued pretraining
How much data (ultrasound images) is typically needed to effectively continue pretraining DINOv2? 

2. Using self-pretrained DINOv2 weights
After training with `lightly_train.pretrain`, I get `last.ckpt` and `exported_last.pt`. What is the correct way to load and use these weights with the model for inference or feature extraction? It seems that `lightly_train.embed` outputs embeddings of shape `[N, 384]`, which looks like CLS features only. Is this expected? How can I extract patch features?

Thanks!

**Comment by liopeer** [comment]
Hey @wudier483,

1. In general, the more the better, but we generally say that curated datasets of at least 100k+ images are a really good starting point. For curating datasets I would check out [LightlyStudio](https://github.com/lightly-ai/lightly-studio).
2. Beyond using the `lightly_train.embed` command there is currently not super convenient way of loading the model weights for inference, but the following should work:
```python
from lightly_train._models.dinov2_vit.dinov2_vit_package import DINOV2_VIT_PACKAGE

model = DINOV2_VIT_PACKAGE.get_model("vits14") # model name without package prefix; same as in pretraining
model.load_state_dict(torch.load("path/to/exported_model.pt", weights_only=True))
```
I thought this should also be logged to your stdout at the end of training. Is this not the case?

**Comment by wudier483** [comment]
Hello, thank you for your reply.
﻿
I have collected around 300,000 thyroid ultrasound images, although many of them are quite similar. I am training with a batch size of 100, gradient accumulation of 4, and I set the total number of epochs to 200.
﻿
At around epoch 22, my loss curves look like the following:

<img width="1398" height="546" alt="Image" src="https://github.com/user-attachments/assets/60bcb924-e53d-4080-a8d0-d7e3b9145ed1" />

﻿
Aside from the iBOT loss increasing, the other losses are still slowly decreasing. I would like to ask:
﻿
Does this indicate that the model has already converged?
Is 22 epochs too early for this kind of training?
The behavior seems quite different from what is described in the official documentation— is this normal?

In addition, I noticed that during the early stage of training, all losses sometimes increase sharply for a short period.
﻿
What could be the reason for this sudden spike at the beginning of training?

**Comment by liopeer** [comment]
Hey, thanks for the detailed info on the training!

With an effective batch size of 400 a single epoch would be 750 steps, so your 200 epochs would be 150k steps which is very close to our default value (125k steps). Your losses at epoch 22 don't look suspicious to me and I would definitely keep going. If you go for very long schedules you will see collapse of the spatial features, similar to how it's really nicely shown in this issue #692 . For a loss curve of a run that probably also lead to a deterioration of the spatial features you can refer to #481 .

I would highly recommend saving intermediate checkpoints through the settings described here: https://docs.lightly.ai/train/stable/settings/pretrain_settings.html#checkpoint-saving . You can then also get a more qualitative impression of your backbone with e.g. the help of this notebook #865.

Small spikes in DINOv2 training are not a big worry, but yours are definitely a bit bigger than what I would usually experience. I don't have a direct answer, but getting insights into intermediate backbones and fine-tuning quality would be helpful.

Two questions: 
- Which model size are you training?
- Just for confirmation: You are continuing from DINOv2 weights like the issue title suggests, right?

**Comment by wudier483** [comment]
Thank you for your reply. Below is part of my training configuration. I adopt the model dinov2_vits14_reg, load its pre-trained weights, and conduct further pre-training on it using ultrasound images.

lightly_train.pretrain(
        out="out/my_experiment", 
        data="data",
        model="dinov2/vits14",
        method="dinov2",
        overwrite=True,
        batch_size=100,
        epochs=200,
        transform_args=transform_args,
        trainer_args = trainer_args,
        method_args = method_args,
        loggers={"wandb": {"project": "dino"}}
    )

**Comment by liopeer** [comment]
Okay, thanks a lot. Just one thing, as said: I recommend saving intermediate checkpoints! Let me know once you have more results, really interested in this!

**Comment by wudier483** [comment]
This is the feature map generated from my checkpoint at 22 epochs. What do you think of it?
<img width="960" height="478" alt="Image" src="https://github.com/user-attachments/assets/4a25990d-64b4-43bb-a627-9b16c8ae131d" />

<img width="952" height="481" alt="Image" src="https://github.com/user-attachments/assets/ce36bfa7-f116-4182-944c-c23529333a50" />

<img width="967" height="487" alt="Image" src="https://github.com/user-attachments/assets/53b15e71-c99a-40eb-8db4-3cdd52623796" />

<img width="965" height="486" alt="Image" src="https://github.com/user-attachments/assets/9472c91c-b008-4d8d-af88-75aa918296a3" />

<img width="957" height="487" alt="Image" src="https://github.com/user-attachments/assets/79b0f7f2-8dd0-4c06-93ce-313bf6384e77" />

**Comment by liopeer** [comment]
Looks like it's learning some semantic regions, but not strongly accentuated so far. What is your downstream task that you would like to use the model for later on?

**Comment by wudier483** [comment]
I plan to use this for a downstream autonomous ultrasound policy model. Compared to the official DINOv2 pretrained weights, I find that the features obtained after training on ultrasound data are slightly better, but the improvement is not substantial.

The comparison is shown as follows: the first image is the original ultrasound image, the second is the feature map after pretraining on ultrasound data, and the third is the feature map from the official DINOv2 weights. When using the original weights, there is clearly a lot of noise. After training, the noise is reduced, but the internal texture becomes somewhat blurry and still lacks sufficient clarity.
Do you have any suggestions for improving the training process?

<img width="248" height="254" alt="Image" src="https://github.com/user-attachments/assets/6ec1a282-d94d-4ab4-ad0c-51ef2db785b9" />
<img width="243" height="257" alt="Image" src="https://github.com/user-attachments/assets/28006261-38e6-47d1-bb32-4175c147c3f8" />
<img width="246" height="254" alt="Image" src="https://github.com/user-attachments/assets/cac9635a-3969-44e8-a473-40d03bd8e02a" />

Finally, I believe that pretraining with DINOv3 would likely yield better results. Could you let me know when pretraining for it might become available?

**Comment by liopeer** [comment]
It definitely looks more "appealing" than the official DINOv2 weights, but I am not blown away either since the boundaries look rather arbitrary (I am definitely not an ultrasound expert though).

I would actually try 2 things:
- train DINOv2 from scratch: for domains that are very different from photographic imagery (+ sufficient data) we have sometimes seen improvements from this
- you can already try our DINOv2 post-training method PaKa: https://github.com/lightly-ai/lightly-train/pull/837 . you would just need to install from that branch and start from the DINOv2 weights that you just trained. PaKa is similar to DINOv3 (but in our internal research better) in that it ensures that there is no collapse of local features

But I will definitely let you know once we have DINOv3!

**Comment by wudier483** [comment]
Thank you very much for your explanation. I will keep an eye on the PaKa method. For now, I plan to first adjust the training parameters to achieve better loss convergence.

In the next stage, I need to integrate the trained DINOv2 parameters (the .pt file) into my policy model. What would be a more elegant way to import the DINOv2 model and load the pretrained weights into the model? During policy training, the DINOv2 backbone may also be further fine-tuned.

**Comment by liopeer** [comment]
The best way to load it is through:
```python
from lightly_train._models.dinov2_vit.dinov2_vit_package import DINOV2_VIT_PACKAGE

model = DINOV2_VIT_PACKAGE.load_model("vits14")
model.load_state_dict("out/exported/exported_last.pt")
```

**Comment by wudier483** [comment]
Thanks, why did the paka branch disappear?

**Comment by gabrielfruet** [comment]
Hi @wudier483 , it is merged into main :) 

[Add: dinov31 DINOv2 post-train method with PaKA dense-relational loss (](https://github.com/lightly-ai/lightly-train/commit/be3a7c9239ef3d1554025d91cb50ba91d52e1701)
